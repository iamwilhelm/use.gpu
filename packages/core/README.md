# @use-gpu/core

```sh
npm install --save @use-gpu/core
```

```sh
yarn add @use-gpu/core
```

# Use.GPU - Core

This package contains various (non-reactive) helper functions for driving WebGPU.

- Animation easing
- Attribute bindings
- Binding generation
- Buffers and textures
- Data aggregation
- Data serialization
- Render pipelines
- MIP resizing
- Storage and uniforms
- Texture atlas

## Colofon

Made by [Steven Wittens](https://acko.net). Part of `@use-gpu`.

