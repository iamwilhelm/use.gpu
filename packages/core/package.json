{
  "name": "@use-gpu/core",
  "version": "0.0.1",
  "main": "src/index.js",
  "author": "Steven Wittens <steven@acko.net>",
  "license": "MIT",
  "private": true,
  "typesVersions": {
    "*": { "types": ["src/types"] }
  },
  "exports": {
    ".": "./src/index.ts",
    "./types": "./src/types.ts"
  },
  "scripts": {
    "test": "jest --watch",
    "lint": "eslint src --ext ts",
    "coverage": "echo 'jest --collect-coverage' disabled",
    "types": "tsc --noEmit",
    "types:watch": "yarn run types -- --watch",

    "build": "yarn run build:types && yarn run build:js && yarn run build:pkg",
    "build:types": "tsc --emitDeclarationOnly",
    "build:js": "babel src --out-dir ../../build/packages/core/src --extensions \".ts\" --source-maps inline",
    "build:pkg": "cp package.json ../../build/packages/core"
  }
}
