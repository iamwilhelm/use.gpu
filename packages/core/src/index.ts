export * from './aggregate';
export * from './atlas';
export * from './attribute';
export * from './binding';
export * from './buffer';
export * from './bytes';
export * from './color';
export * from './constants';
export * from './data';
export * from './depth';
export * from './mip';
export * from './picking';
export * from './pipeline';
export * from './storage';
export * from './texture';
export * from './tuple';
export * from './uniform';
export * from './view';

export * from './types';
