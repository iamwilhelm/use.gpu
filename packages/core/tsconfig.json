{
  "extends": "../../tsconfig.pkg.json",
  "references": [
    { "path": "../state" }
  ],
  "compilerOptions": {
    "rootDir": "..",
    "baseUrl": "src",
    "outDir": "../../build/ts",
    "tsBuildInfoFile": "./tsconfig.tsbuildinfo"
  },
  "include": ["src/**/*.ts", "types.ts"]
}