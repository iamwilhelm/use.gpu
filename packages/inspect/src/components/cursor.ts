import { useCallback, useMemo, useState } from 'react';
import { Update } from '@use-gpu/state/types';

import { patch, $set } from '@use-gpu/state';

export type Cursor<T> = [T, Updater<T>];
export type Updater<T> = (u: Update<T>) => void;

type Key = string | number;
type RefineCursor<T> = (cursor: Cursor<any>) => (...keys: Key[]) => Cursor<T>

export const refineCursor = <T>(
  cursor: Cursor<any>,
  defaults?: Record<string, any>,
) => <S = T>(...keys: Key[]): Cursor<S> => {
  let [v, u] = cursor;
  let d = defaults;

  for (const k of keys) {
    let updater = u;

    d = d != null ? d[k] : undefined;
    v = v != null ? v[k] : undefined;

    if (v === undefined && d !== undefined) {
      let once = true;

      v = d;
      u = (update: any) => {
        if (once) updater({ [k]: $set(d) });
        updater({ [k]: update });
        once = false;
      };
    }
    else {
      u = (update: any) => updater({ [k]: update });
    }
  }
  return [v, u];
};

export const useRefineCursor = <T>(cursor: Cursor<any>) => <S = T>(...keys: Key[]) => (
  useMemo(() => refineCursor(cursor)<S>(...keys), [cursor, ...keys])
);

export const useUpdateState = <T>(initialState: T | (() => T)): Cursor<T> => {
  const [state, setState] = useState(initialState);

  const updateState = useCallback((update: Update<T>) =>
    setState((s) => patch(s, update)),
    []);

  // eslint-disable-next-line
  return useMemo(() => [state, updateState] as Cursor<T>, [state]);
}
