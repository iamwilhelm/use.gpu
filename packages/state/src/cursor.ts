import { Cursor, Pair, Update, UpdateKey } from './types';
import { $set } from './patch';

export type CursorEntry<T> = [Cursor<T> | null, CursorMap<T>];
export type CursorMap<T> = Map<string | symbol, CursorEntry<any>>;

const NO_ENTRY = [null, null];

/** Make a `Cursor<T>` from a `[value, updater]` pair.

- Also returns mutable `map` of cached cursors. This is filled in whenever `cursor.foo.â€¦` is accessed.
- When creating a replacement cursor, pass `map` as `keep` to preserve child cursors for unchanged parts.
*/
export const makeCursor = <T>(
  pair: Pair<T>,
  defaults?: T,
  keep?: CursorMap<T> | null,
): [Cursor<T>, CursorMap<T>] => {
  const d = defaults as any;

  const fn = () => pair;
  const map: CursorMap<T> = new Map();

  const cursor = (
    new Proxy(fn, {
      get(target, prop, receiver) {
        const [cursor, keep] = map.get(prop) ?? NO_ENTRY;
        if (cursor) return cursor;

        const child = refinePair(pair, prop as any, d);
        const entry = makeCursor(child, d?.[prop], keep);

        map.set(prop, entry);
        return entry[0];
      },
    }) as Cursor<T>
  );

  if (keep) {
    if (pair[0]) for (const k of keep.keys()) {
      const [newValue] = refinePair(pair, k as any, d);

      const entry = keep.get(k)!;
      const [cursor, m] = entry;
      if (cursor) {
        const [oldValue] = cursor();
        if (oldValue === newValue) {
          map.set(k, entry);
          continue;
        }
      }
      if (newValue) map.set(k, [null, m]);
    }
  }

  return [cursor, map];
};

/**
Refine a getter/updater pair to traverse its `value` along `keys` and produce new cursors.

If `defaults` are supplied, the cursor will merge them in.
When the first update is made, it will dispatch a $set operation first
to permanently insert them, and allow a clean patch to be made.

The first part of the higher-order function can be called once and then reused for different paths.
*/
export const refinePair = <T, K extends keyof T>(
  pair: Pair<T>,
  key: K,
  defaults?: Record<K, any>,
): Pair<T[K]> => {
  let [value, updater] = pair;

  let d = defaults != null ? defaults[key]       : undefined;
  let v = value    != null ? (value as any)[key] : undefined;
  let u;

  // Fill in default
  if (v === undefined && d !== undefined) {
    let once = true;

    v = d;
    u = (update: any) => {
      if (once) updater({ [key]: $set(d) } as Update<T>);
      updater({ [key]: update } as Update<T>);
      once = false;
    };
  }
  // Existing value
  else {
    u = (update: any) => updater({ [key]: update } as Update<T>);
  }

  return [v, u];
};
