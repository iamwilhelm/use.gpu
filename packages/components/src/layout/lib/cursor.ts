import { Alignment } from '../types';

import { makeTuples } from '@use-gpu/core';
import { evaluateAnchor } from '../parse';

type Reduce = (start: number, end: number, gap: number, lead: number, count: number, cross: number, ascent: number, descent: number, index: number) => void;

type LayoutCursor = {
  push: (advance: number, trim: number, hard: number, cross?: number, base?: number, descent?: number) => void,
  flush: (x: number) => void,
  gather: (reduce: Reduce) => Float32Array,
};

// Layout cursor for putting inline items on lines, with line wrapping, alignment and justification.
export const makeLayoutCursor = (
  max: number,
  align: Alignment,
): LayoutCursor => {

  let spanCount = 0;
  let spanAdvance = 0;
  let spanTrim = 0;
  let spanCross = 0;
  let spanBase = 0;
  let spanDescent = 0;

  let start: number = 0;
  let end: number = 0;
  
  let chunkAdvance = 0;
  let chunkIndex = 0;
  let chunkCross = 0;

  let rows: number[] = [];
  let sizes: number[] = [];
  let index = 0;

  const push = (advance: number, trim: number, hard: number, cross: number = 0, base: number = 0, descent: number = 0) => {
    if (max > 0 && (spanAdvance + advance - trim > max)) {
      flush(0);
      start = index;
    }

    end = index + 1;

    if (trim) {
      spanCount++;
      spanTrim = trim;
    }
    spanAdvance += advance;
    spanCross = Math.max(spanCross, cross);
    spanBase = Math.max(spanBase, base);
    spanDescent = Math.max(spanDescent, descent);

    if (hard) {
      flush(hard);
      start = end;
    }

    index++;
  };

  const flush = (hard: number) => {
    const n = end > start;
    if (n) {
      const spanSize = spanAdvance - spanTrim;
      chunkAdvance = Math.max(chunkAdvance, spanSize);

      rows.push(start, end, hard, spanSize, spanCount, spanCross, spanBase, spanDescent, chunkIndex);
      chunkCross += spanCross;
    }

    spanCount = 0;
    spanAdvance = 0;
    spanTrim = 0;
    spanCross = 0;
    spanBase = 0;
    spanDescent = 0;

    if (hard === 2) {
      chunkIndex++;
      sizes.push(chunkAdvance, chunkCross);

      chunkAdvance = 0;
      chunkCross = 0;
    }
  };
  
  const gather = (reduce: Reduce) => {
    flush(2);

    const s = makeTuples(sizes, 2);
    const r = makeTuples(rows, 9);
    r.iterate((start: number, end: number, hard: number, advance: number, count: number, cross: number, base: number, descent: number, index: number) => {
      const slack = (max || s.get(index, 0)) - advance;
      const [gap, lead] = getAlignmentSpacing(slack, count, !!hard, align);
      reduce(start, end, gap, lead, count, cross, base, descent, index);
    });

    return new Float32Array(sizes);
  };

  return { push, flush, gather };
}

export const getAlignmentSpacing = (
  slack: number,
  n: number,
  hard: boolean,
  align: Alignment,
) => {
  let gap = 0;
  let lead = 0;

  const isJustifyStart  = align === 'justify-start';
  const isJustifyCenter = align === 'justify-center';
  const isJustifyEnd    = align === 'justify-end';

  const isJustify = align === 'justify' ||
                    ((isJustifyStart || isJustifyCenter || isJustifyEnd) && !hard);
  const isBetween = align === 'between';
  const isEvenly  = align === 'evenly';

  if (slack > 0) {
    if (isEvenly || isBetween || isJustify) {
      if (n === 1) {
        lead = slack / 2;
      }
      else if (isEvenly) {
        gap = Math.max(0, slack / (n + 1));
        lead = gap;
      }
      else if (isBetween) {
        gap = Math.max(0, slack / n);
        lead = gap / 2;
      }
      else if (isJustify) {
        gap = Math.max(0, slack / Math.max(1, n - 1));
      }
    }
    else {
      lead = evaluateAnchor(align) * slack;
    }
  }

  return [gap, lead];
};
