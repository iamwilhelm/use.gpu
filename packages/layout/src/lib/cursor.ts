import type { Point } from '@use-gpu/core';
import type { Alignment } from '../types';

import { makeTuples } from '@use-gpu/core';
import { getAlignmentSpacing } from './util';

type InlineReduce = (
  start: number,
  end: number,
  gap: number,
  lead: number,
  count: number,
  cross: number,
  ascent: number,
  descent: number,
  xHeight: number,
  index: number,
) => void;

type InlineCursor = {
  push: (
    advance: number,
    trim: number,
    hard: number,
    cross: number,
    base: number,
    descent: number,
    xHeight: number,
  ) => void;
  flush: (x: number) => void,
  gather: (reduce: InlineReduce) => Float32Array,
};

// Layout cursor for putting inline items on lines, with line wrapping, alignment and justification.
export const makeInlineCursor = (
  max: number,
  align: Alignment,
  even?: boolean,
): InlineCursor => {

  let spanCount = 0;
  let spanAdvance = 0;
  let spanTrim = 0;
  let spanCross = 0;
  let spanAscent = 0;
  let spanDescent = 0;
  let spanXHeight = 0;

  let start: number = 0;
  let end: number = 0;
  
  let chunkAdvance = 0;
  let chunkIndex = 0;
  let chunkCross = 0;

  let rows: number[] = [];
  let sizes: number[] = [];
  let index = 0;

  const push = (
    advance: number,
    trim: number,
    hard: number,
    cross: number,
    ascent: number,
    descent: number,
    xHeight: number,
  ) => {
    if (max > 0 && (spanAdvance + advance - trim > max)) {
      if (!spanTrim) spanCount++;

      flush(0);
      start = index;
    }

    end = index + 1;

    if (hard) {
      spanCount++;
    }
    else if (trim) {
      spanCount++;
    }

    spanAdvance += advance;
    spanCross = Math.max(spanCross, cross);
    spanAscent = Math.max(spanAscent, ascent);
    spanDescent = Math.max(spanDescent, descent);
    spanXHeight = Math.max(spanXHeight, xHeight);
    spanTrim = trim;

    if (hard) {
      flush(hard);
      start = end;
    }

    index++;
  };

  const flush = (hard: number) => {
    const n = end > start;
    if (n) {
      const spanSize = spanAdvance - spanTrim;
      chunkAdvance = Math.max(chunkAdvance, spanSize);

      rows.push(start, end, hard, spanSize, spanCount, spanCross, spanAscent, spanDescent, spanXHeight, chunkIndex);
      chunkCross += spanCross;
    }

    spanCount = 0;
    spanAdvance = 0;
    spanTrim = 0;
    spanCross = 0;
    spanAscent = 0;
    spanDescent = 0;

    if (hard === 2) {
      chunkIndex++;
      sizes.push(chunkAdvance, chunkCross);

      chunkAdvance = 0;
      chunkCross = 0;
    }
  };
  
  const gather = (reduce: InlineReduce) => {
    spanCount++;
    flush(2);

    const s = makeTuples(sizes, 2);
    const r = makeTuples(rows, 10);
    r.iterate((
      start: number,
      end: number,
      hard: number,
      advance: number,
      count: number,
      cross: number,
      ascent: number,
      descent: number,
      xHeight: number,
      index: number,
    ) => {
      let slack;
      if (even) {
        slack = (Math.floor((max || s.get(index, 0)) / 2) - Math.floor(advance / 2)) * 2;
      } else {
        slack = (max || s.get(index, 0)) - advance;
      }
      const [gap, lead] = getAlignmentSpacing(slack, count, !!hard, align);
      reduce(start, end, gap, lead, count, cross, ascent, descent, xHeight, index);
    });

    return new Float32Array(sizes);
  };

  return { push, flush, gather };
}

type FlexReduce = (
  sizes: number[],
  start: number,
  end: number,
  gap: number,
  lead: number,
) => void;

type FlexCursor = {
  push: (
    main: number,
    margin: number,
    gap: number,
    grow: number,
    shrink: number,
  ) => void;
  flush: (x: number) => void,
  gather: (reduce: FlexReduce) => number,
};

// Layout cursor for putting boxes on flex lines, with optional wrapping, grow/shrink, alignment and justification.
export const makeFlexCursor = (
  into: number | null,
  align: Alignment,
  wrap?: boolean,
): FlexCursor => {

  let spanMain = 0;
  let spanTrim = 0;
  let chunkMain = 0;

  let start: number = 0;
  let end: number = 0;

  let rows: number[] = [];
  let sizes: number[] = [];
  let grows: number[] = [];
  let shrinks: number[] = [];
  let index = 0;

  const push = (
    main: number,
    margin: number,
    gap: number,
    grow: number,
    shrink: number,
  ) => {
    if (into != null && wrap && (spanMain + main + margin > into)) {
      flush();
      start = index;
    }

    end = index + 1;

    spanMain += main + margin + gap;
    spanTrim  = gap;

    sizes.push(main);
    grows.push(grow);
    shrinks.push(shrink);

    index++;
  };

  const flush = () => {
    const n = end > start;
    if (n) {
      const spanSize = spanMain - spanTrim;
      chunkMain = Math.max(chunkMain, spanSize);
      if (Number.isNaN(chunkMain)) debugger;

      rows.push(start, end, spanSize);
    }

    spanMain = 0;
    spanTrim = 0;
  };
  
  const gather = (reduce: FlexReduce): number => {
    flush();

    const r = makeTuples(rows, 3);
    const l = r.length - 1;

    r.iterate((
      start: number,
      end: number,
      main: number,
      index: number,
    ) => {
      let slack = (into ?? chunkMain) - main;
      
      if (slack > 0 && growRow(slack, grows, sizes, start, end)) slack = 0;
      if (slack < 0 && shrinkRow(slack, shrinks, sizes, start, end)) slack = 0;
      
      const [gap, lead] = getAlignmentSpacing(slack, end - start, index === l, align);

      reduce(sizes, start, end, gap, lead);
    });
    
    return into ?? chunkMain;
  };

  return { push, flush, gather };
}


// Grow all applicable blocks in a row to add extra slack.
export const growRow = (slack: number, grow: number[], sizes: number[], start: number, end: number) => {
  let weight = 0;
  for (let i = start; i < end; ++i) if (grow[i] > 0) weight += grow[i];

  if (weight > 0) {
    for (let i = start; i < end; ++i) if (grow[i] > 0) {
      sizes[i] += slack * grow[i] / weight;
    }
    return true;
  }
  return false;
}

// Shrink all applicable blocks in a row to remove excess slack.
export const shrinkRow = (slack: number, shrink: number[], sizes: number[], start: number, end: number): boolean => {
  let weight = 0;
  for (let i = start; i < end; ++i) if (shrink[i]) weight += shrink[i] * sizes[i];

  if (weight > 0) {
    let negative = 0;
    for (let i = start; i < end; ++i) if (shrink[i] > 0 && sizes[i]) {
      sizes[i] += slack * shrink[i] * sizes[i] / weight;
      if (sizes[i] < 0) {
        negative += sizes[i];
        sizes[i] = 0;
      }
    }
    if (negative) {
      shrinkRow(negative, shrink, sizes, start, end);
    }
    return true;
  }
  return false;
}

