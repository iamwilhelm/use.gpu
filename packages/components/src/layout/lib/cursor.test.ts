import { makeTuples } from '@use-gpu/core';
import { makeLayoutCursor } from './cursor';

describe('layout cursor', () => {
  
  const lineHeight = 30;
  
  it('makes centered rows with hard breaks and multiple items', () => {
    
    const cursor = makeLayoutCursor(0, 'center');
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 1, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 2, lineHeight);
    cursor.push(200, 10, 0, lineHeight);
    cursor.push(100, 10, 1, lineHeight);
    cursor.push(100, 10, 1, lineHeight);

    const out: number[][] = [];
    const blocks = cursor.gather((start, end, gap, lead, count, cross, base, index) => out.push([start, end, gap, lead, index]));
    
    expect(out).toMatchSnapshot();
    expect(blocks).toMatchSnapshot();
  });

  it('makes centered rows with soft-wrapped breaks', () => {
    
    const cursor = makeLayoutCursor(390, 'center');
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(200, 10, 0, lineHeight);
    cursor.push(200, 10, 0, lineHeight);
    cursor.push(100, 10, 1, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(300, 10, 0, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 1, lineHeight);
    cursor.push(100, 10, 1, lineHeight);

    const out: number[][] = [];
    const blocks = cursor.gather((start, end, gap, lead, count, cross, base, index) => out.push([start, end, gap, lead, index]));
    
    expect(out).toMatchSnapshot();
    expect(blocks).toMatchSnapshot();
  });
  
  it('makes justified rows with soft-wrapped breaks', () => {
    
    const cursor = makeLayoutCursor(390, 'justify');
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(200, 10, 0, lineHeight);
    cursor.push(200, 10, 0, lineHeight);
    cursor.push(100, 10, 1, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(300, 10, 0, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 1, lineHeight);
    cursor.push(100, 10, 1, lineHeight);

    const out: number[][] = [];
    const blocks = cursor.gather((start, end, gap, lead, count, cross, base, index) => out.push([start, end, gap, lead, index]));
    
    expect(out).toMatchSnapshot();
    expect(blocks).toMatchSnapshot();
  });
  
  it('makes right-justified rows with soft-wrapped breaks', () => {
    
    const cursor = makeLayoutCursor(390, 'justify-end');
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(200, 10, 0, lineHeight);
    cursor.push(200, 10, 0, lineHeight);
    cursor.push(100, 10, 1, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(300, 10, 0, lineHeight);
    cursor.push(100, 10, 0, lineHeight);
    cursor.push(100, 10, 1, lineHeight);
    cursor.push(100, 10, 1, lineHeight);

    const out: number[][] = [];
    const blocks = cursor.gather((start, end, gap, lead, count, cross, base, index) => out.push([start, end, gap, lead, index]));
    
    expect(out).toMatchSnapshot();
    expect(blocks).toMatchSnapshot();
  });
});
