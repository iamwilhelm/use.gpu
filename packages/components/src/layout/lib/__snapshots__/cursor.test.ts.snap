// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout cursor makes centered rows with hard breaks and multiple items 1`] = `
Array [
  Array [
    0,
    4,
    0,
    0,
    0,
  ],
  Array [
    4,
    7,
    0,
    50,
    0,
  ],
  Array [
    7,
    9,
    0,
    0,
    1,
  ],
  Array [
    9,
    10,
    0,
    100,
    1,
  ],
]
`;

exports[`layout cursor makes centered rows with hard breaks and multiple items 2`] = `
Float32Array [
  390,
  60,
  290,
  60,
]
`;

exports[`layout cursor makes centered rows with soft-wrapped breaks 1`] = `
Array [
  Array [
    0,
    3,
    0,
    0,
    0,
  ],
  Array [
    3,
    5,
    0,
    50,
    0,
  ],
  Array [
    5,
    7,
    0,
    0,
    0,
  ],
  Array [
    7,
    9,
    0,
    100,
    0,
  ],
  Array [
    9,
    10,
    0,
    150,
    0,
  ],
]
`;

exports[`layout cursor makes centered rows with soft-wrapped breaks 2`] = `
Float32Array [
  390,
  150,
]
`;

exports[`layout cursor makes justified rows with soft-wrapped breaks 1`] = `
Array [
  Array [
    0,
    3,
    0,
    0,
    0,
  ],
  Array [
    3,
    5,
    100,
    0,
    0,
  ],
  Array [
    5,
    7,
    0,
    0,
    0,
  ],
  Array [
    7,
    9,
    200,
    0,
    0,
  ],
  Array [
    9,
    10,
    0,
    150,
    0,
  ],
]
`;

exports[`layout cursor makes justified rows with soft-wrapped breaks 2`] = `
Float32Array [
  390,
  150,
]
`;

exports[`layout cursor makes right-justified rows with soft-wrapped breaks 1`] = `
Array [
  Array [
    0,
    3,
    0,
    0,
    0,
  ],
  Array [
    3,
    5,
    0,
    100,
    0,
  ],
  Array [
    5,
    7,
    0,
    0,
    0,
  ],
  Array [
    7,
    9,
    0,
    200,
    0,
  ],
  Array [
    9,
    10,
    0,
    300,
    0,
  ],
]
`;

exports[`layout cursor makes right-justified rows with soft-wrapped breaks 2`] = `
Float32Array [
  390,
  150,
]
`;
